### ISSUE TITLE
[BUG] matx::matmul can't be performed on tensors created from user-defined pointers

### ISSUE THREAD

**Issue by ZJUGuoShuai** [NVIDIA/MatX#126]
**Describe the bug**
I'm trying to use `matx::matmul` to perform GEMM on two tensors created from device pointers allocated by CUDA APIs. But the code didn't compile for some reasons.

**To Reproduce**
1. Prepare the data buffers on CPU:
```cpp
float data_a[4] = {1.0, 2.0, 3.0, 4.0};
float data_b[4] = {1.0, 2.0, 3.0, 4.0};
```
2. Allocate GPU memory for three 2x2 tensors using `cudaMalloc`:
```cpp
float *pa, *pb, *pc;
CUDA_CHECK(cudaMalloc(&pa, sizeof(float) * 4));
CUDA_CHECK(cudaMalloc(&pb, sizeof(float) * 4));
CUDA_CHECK(cudaMalloc(&pc, sizeof(float) * 4));
```
3. Copy the data to GPU:
```cpp
CUDA_CHECK(cudaMemcpy(pa, data_a, sizeof(float) * 4, cudaMemcpyHostToDevice));
CUDA_CHECK(cudaMemcpy(pb, data_b, sizeof(float) * 4, cudaMemcpyHostToDevice));
```
4. Create 3 tensors from these pointers:
```cpp
auto ta = matx::make_tensor<float, 2, matx::non_owning>(pa, {2, 2});
auto tb = matx::make_tensor<float, 2, matx::non_owning>(pb, {2, 2});
auto tc = matx::make_tensor<float, 2, matx::non_owning>(pc, {2, 2});
```
5. Perform `matmul`:
```cpp
matx::matmul(tc, ta, tb);
tc.Print();
```
6. Compiling failed:
```
[ 98%] Building CUDA object test/operator_tests/CMakeFiles/embedding.dir/embedding.cu.o
/root/gs/nexus3/3rdparty/MatX/include/matx_tensor_desc.h(230): error: no operator "=" matches these operands
            operand types are: std::array<long long, 2UL> = const matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>
          detected during:
            instantiation of "void matx::tensor_desc_t<ShapeType, StrideType, RANK>::InitFromShape(S2 &&) [with ShapeType=std::array<long long, 2UL>, StrideType=std::array<long long, 2UL>, RANK=2, S2=const matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>> &]" 
(112): here
            instantiation of "matx::tensor_desc_t<ShapeType, StrideType, RANK>::tensor_desc_t(S2 &&) [with ShapeType=std::array<long long, 2UL>, StrideType=std::array<long long, 2UL>, RANK=2, S2=const matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>> &, <unnamed>=true]" 
/root/gs/nexus3/3rdparty/MatX/include/matx_matmul.h(225): here
            instantiation of "matx::detail::MatMulParams_t matx::detail::matxMatMulHandle_t<TensorTypeC, TensorTypeA, TensorTypeB, PROV>::GetGemmParams(TensorTypeC &, const TensorTypeA &, const TensorTypeB &) [with TensorTypeC=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeA=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeB=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, PROV=(matx::MatXMatMulProvider_t)1]" 
/root/gs/nexus3/3rdparty/MatX/include/matx_matmul.h(916): here
            instantiation of "void matx::matmul(TensorTypeC &, const TensorTypeA &, const TensorTypeB &, cudaStream_t, float, float) [with TensorTypeC=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeA=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeB=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, PROV=(matx::MatXMatMulProvider_t)1]" 
/root/gs/nexus3/test/operator_tests/embedding.cu(36): here

/root/gs/nexus3/3rdparty/MatX/include/matx_tensor_desc.h(233): error: class "matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>" has no member "begin"
          detected during:
            instantiation of "void matx::tensor_desc_t<ShapeType, StrideType, RANK>::InitFromShape(S2 &&) [with ShapeType=std::array<long long, 2UL>, StrideType=std::array<long long, 2UL>, RANK=2, S2=const matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>> &]" 
(112): here
            instantiation of "matx::tensor_desc_t<ShapeType, StrideType, RANK>::tensor_desc_t(S2 &&) [with ShapeType=std::array<long long, 2UL>, StrideType=std::array<long long, 2UL>, RANK=2, S2=const matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>> &, <unnamed>=true]" 
/root/gs/nexus3/3rdparty/MatX/include/matx_matmul.h(225): here
            instantiation of "matx::detail::MatMulParams_t matx::detail::matxMatMulHandle_t<TensorTypeC, TensorTypeA, TensorTypeB, PROV>::GetGemmParams(TensorTypeC &, const TensorTypeA &, const TensorTypeB &) [with TensorTypeC=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeA=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeB=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, PROV=(matx::MatXMatMulProvider_t)1]" 
/root/gs/nexus3/3rdparty/MatX/include/matx_matmul.h(916): here
            instantiation of "void matx::matmul(TensorTypeC &, const TensorTypeA &, const TensorTypeB &, cudaStream_t, float, float) [with TensorTypeC=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeA=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeB=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, PROV=(matx::MatXMatMulProvider_t)1]" 
/root/gs/nexus3/test/operator_tests/embedding.cu(36): here

/root/gs/nexus3/3rdparty/MatX/include/matx_tensor_desc.h(230): error: no operator "=" matches these operands
            operand types are: std::array<long long, 2UL> = matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>
          detected during:
            instantiation of "void matx::tensor_desc_t<ShapeType, StrideType, RANK>::InitFromShape(S2 &&) [with ShapeType=std::array<long long, 2UL>, StrideType=std::array<long long, 2UL>, RANK=2, S2=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>> &]" 
(112): here
            instantiation of "matx::tensor_desc_t<ShapeType, StrideType, RANK>::tensor_desc_t(S2 &&) [with ShapeType=std::array<long long, 2UL>, StrideType=std::array<long long, 2UL>, RANK=2, S2=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>> &, <unnamed>=true]" 
/root/gs/nexus3/3rdparty/MatX/include/matx_matmul.h(227): here
            instantiation of "matx::detail::MatMulParams_t matx::detail::matxMatMulHandle_t<TensorTypeC, TensorTypeA, TensorTypeB, PROV>::GetGemmParams(TensorTypeC &, const TensorTypeA &, const TensorTypeB &) [with TensorTypeC=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeA=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeB=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, PROV=(matx::MatXMatMulProvider_t)1]" 
/root/gs/nexus3/3rdparty/MatX/include/matx_matmul.h(916): here
            instantiation of "void matx::matmul(TensorTypeC &, const TensorTypeA &, const TensorTypeB &, cudaStream_t, float, float) [with TensorTypeC=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeA=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeB=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, PROV=(matx::MatXMatMulProvider_t)1]" 
/root/gs/nexus3/test/operator_tests/embedding.cu(36): here

/root/gs/nexus3/3rdparty/MatX/include/matx_tensor_desc.h(233): error: class "matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>" has no member "begin"
          detected during:
            instantiation of "void matx::tensor_desc_t<ShapeType, StrideType, RANK>::InitFromShape(S2 &&) [with ShapeType=std::array<long long, 2UL>, StrideType=std::array<long long, 2UL>, RANK=2, S2=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>> &]" 
(112): here
            instantiation of "matx::tensor_desc_t<ShapeType, StrideType, RANK>::tensor_desc_t(S2 &&) [with ShapeType=std::array<long long, 2UL>, StrideType=std::array<long long, 2UL>, RANK=2, S2=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>> &, <unnamed>=true]" 
/root/gs/nexus3/3rdparty/MatX/include/matx_matmul.h(227): here
            instantiation of "matx::detail::MatMulParams_t matx::detail::matxMatMulHandle_t<TensorTypeC, TensorTypeA, TensorTypeB, PROV>::GetGemmParams(TensorTypeC &, const TensorTypeA &, const TensorTypeB &) [with TensorTypeC=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeA=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeB=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, PROV=(matx::MatXMatMulProvider_t)1]" 
/root/gs/nexus3/3rdparty/MatX/include/matx_matmul.h(916): here
            instantiation of "void matx::matmul(TensorTypeC &, const TensorTypeA &, const TensorTypeB &, cudaStream_t, float, float) [with TensorTypeC=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeA=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeB=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, PROV=(matx::MatXMatMulProvider_t)1]" 
/root/gs/nexus3/test/operator_tests/embedding.cu(36): here

/root/gs/nexus3/3rdparty/MatX/include/matx_matmul.h(236): error: no suitable user-defined conversion from "matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>" to "const matx::tensor_t<float, 2, matx::DefaultStorage<float>, matx::DefaultDescriptor<2>>" exists
          detected during:
            instantiation of "matx::detail::MatMulParams_t matx::detail::matxMatMulHandle_t<TensorTypeC, TensorTypeA, TensorTypeB, PROV>::GetGemmParams(TensorTypeC &, const TensorTypeA &, const TensorTypeB &) [with TensorTypeC=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeA=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeB=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, PROV=(matx::MatXMatMulProvider_t)1]" 
(916): here
            instantiation of "void matx::matmul(TensorTypeC &, const TensorTypeA &, const TensorTypeB &, cudaStream_t, float, float) [with TensorTypeC=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeA=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeB=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, PROV=(matx::MatXMatMulProvider_t)1]" 
/root/gs/nexus3/test/operator_tests/embedding.cu(36): here

/root/gs/nexus3/3rdparty/MatX/include/matx_matmul.h(237): error: no suitable user-defined conversion from "matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>" to "const matx::tensor_t<float, 2, matx::DefaultStorage<float>, matx::DefaultDescriptor<2>>" exists
          detected during:
            instantiation of "matx::detail::MatMulParams_t matx::detail::matxMatMulHandle_t<TensorTypeC, TensorTypeA, TensorTypeB, PROV>::GetGemmParams(TensorTypeC &, const TensorTypeA &, const TensorTypeB &) [with TensorTypeC=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeA=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeB=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, PROV=(matx::MatXMatMulProvider_t)1]" 
(916): here
            instantiation of "void matx::matmul(TensorTypeC &, const TensorTypeA &, const TensorTypeB &, cudaStream_t, float, float) [with TensorTypeC=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeA=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeB=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, PROV=(matx::MatXMatMulProvider_t)1]" 
/root/gs/nexus3/test/operator_tests/embedding.cu(36): here

/root/gs/nexus3/3rdparty/MatX/include/matx_matmul.h(238): error: no suitable user-defined conversion from "matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>" to "const matx::tensor_t<float, 2, matx::DefaultStorage<float>, matx::DefaultDescriptor<2>>" exists
          detected during:
            instantiation of "matx::detail::MatMulParams_t matx::detail::matxMatMulHandle_t<TensorTypeC, TensorTypeA, TensorTypeB, PROV>::GetGemmParams(TensorTypeC &, const TensorTypeA &, const TensorTypeB &) [with TensorTypeC=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeA=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeB=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, PROV=(matx::MatXMatMulProvider_t)1]" 
(916): here
            instantiation of "void matx::matmul(TensorTypeC &, const TensorTypeA &, const TensorTypeB &, cudaStream_t, float, float) [with TensorTypeC=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeA=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, TensorTypeB=matx::tensor_t<float, 2, matx::basic_storage<matx::raw_pointer_buffer<float, matx::non_owning, matx::matx_allocator<float>>>, matx::DefaultDescriptor<2>>, PROV=(matx::MatXMatMulProvider_t)1]" 
/root/gs/nexus3/test/operator_tests/embedding.cu(36): here

7 errors detected in the compilation of "/root/gs/nexus3/test/operator_tests/embedding.cu".
make[3]: *** [test/operator_tests/CMakeFiles/embedding.dir/build.make:76: test/operator_tests/CMakeFiles/embedding.dir/embedding.cu.o] Error 1
make[2]: *** [CMakeFiles/Makefile2:1179: test/operator_tests/CMakeFiles/embedding.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:1186: test/operator_tests/CMakeFiles/embedding.dir/rule] Error 2
make: *** [Makefile:533: embedding] Error 2
```

**Expected behavior**
The `matx::matmul` should compile.

**System details (please complete the following information):**
 - OS: Ubuntu 18.04
 - CUDA version: CUDA 11.4
 - g++ version: 9.3.0

**Comment by cliffburdick** [comment]
thanks @ZJUGuoShuai, this will be fixed from bug #110 . we will prioritize it and let you know when it's ready.

**Comment by cliffburdick** [comment]
Hi @ZJUGuoShuai, I pushed a temporary fix for this. It's temporary because the root cause is the way we're creating temporary tensors is not portable in all scenarios (as you found), but in this case, gemms did not need this temporary tensor created. I also added a unit test to make sure this case works, and you can see the code here:

https://github.com/NVIDIA/MatX/blob/main/test/00_transform/MatMul.cu#L96

I'm going to close this one and the other outstanding issue will be fixed in #110

**Comment by ZJUGuoShuai** [comment]
It works now, thanks!
